cliext/oauth.go

```diff
@@ -0,0 +1,244 @@
+	var stateBytes [16]byte
+	if _, err := rand.Read(stateBytes[:]); err != nil {
+		return OAuthToken{}, fmt.Errorf("failed to generate state: %w", err)
+	}
+	state := base64.RawURLEncoding.EncodeToString(stateBytes[:])
+
+	// Generate authorization URL.
+	redirectURL := fmt.Sprintf("http://127.0.0.1:%d/callback", port)
+	cfg := c.createOAuth2Config(redirectURL)
+	authURL := cfg.AuthCodeURL(state, authOpts...)
+
+	// Handle auth URL notification.
+	if c.OnAuthURL != nil {
+		c.OnAuthURL(authURL)
+	} else {
+		fmt.Printf("Opening browser to authorize. If it doesn't open, visit: %s\n", authURL)
+		_ = browser.OpenURL(authURL)
+	}
+
+	// Start HTTP server to handle callback.
+	var once sync.Once
+	resultCh := make(chan oauthCallbackResult, 1)
+	server := &http.Server{
+		Handler: http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			if r.URL.Path != "/callback" {
+				http.NotFound(w, r)
+				return
+			}
+
+			// Use sync.Once to only process the first callback.
+			once.Do(func() {
+				query := r.URL.Query()
+
+				// Check for OAuth error response.
+				if errCode := query.Get("error"); errCode != "" {
+					errDesc := query.Get("error_description")
+					if errDesc == "" {
+						errDesc = errCode
+					}
+					resultCh <- oauthCallbackResult{err: fmt.Errorf("authorization failed: %s", errDesc)}
+					http.Error(w, fmt.Sprintf("Authorization failed: %s", errDesc), http.StatusBadRequest)
+					return
+				}
+
+				// Validate state to prevent CSRF.
+				if query.Get("state") != state {
+					resultCh <- oauthCallbackResult{err: fmt.Errorf("invalid state parameter")}
+					http.Error(w, "Invalid state parameter", http.StatusBadRequest)
+					return
+				}
+
+				// Check for authorization code.
+				code := query.Get("code")
+				if code == "" {
+					resultCh <- oauthCallbackResult{err: fmt.Errorf("missing authorization code")}
+					http.Error(w, "Missing authorization code", http.StatusBadRequest)
+					return
+				}
+
+				resultCh <- oauthCallbackResult{code: code}
+				fmt.Fprint(w, "Authorization successful! You can close this window.")
+			})
+		}),
+	}
+	go server.Serve(listener)
+	defer func() {
+		shutdownCtx, cancel := context.WithTimeout(context.Background(), 2*time.Second)
+		defer cancel()
+		server.Shutdown(shutdownCtx)
+	}()
+
+	// Wait for callback result or context cancellation.
+	var result oauthCallbackResult
+	select {
+	case result = <-resultCh:
+	case <-ctx.Done():
+		return OAuthToken{}, ctx.Err()
+	}
+	if result.err != nil {
+		return OAuthToken{}, result.err
+	}
+
+	// Exchange code for token with PKCE verifier.
+	token, err := cfg.Exchange(ctx, result.code, oauth2.VerifierOption(verifier))
+	if err != nil {
+		return OAuthToken{}, fmt.Errorf("failed to exchange code for token: %w", err)
+	}
+
+	return oauthTokenFromOAuth2(token), nil
+}
+
+// Token returns a valid access token, refreshing if necessary.
+func (c *OAuthClient) Token(ctx context.Context, config *OAuthConfig) (OAuthToken, error) {
```

Comment:
I don't think obtaining a token is part of `cliext`, that is only something needed by client option resolution. Rather the CLI extension library should provide "client connection information" which leverages everything from envconfig, args, etc and gives the CLI extension what it may need to connect (addr, namespace, API key (aka this token sometimes), TLS, etc). Maybe even a `DialClient` since many CLI extensions will likely do data plane connections.

Reply:
To clarify, `DialContext` is not sth we already have right?

So the idea would be to provide a `DialContext` with all the info for a client to connect to data plane (or control plane in the Cloud CLI case), right?

We'll still need the behavior of `Token`, but you're saying rather than exposing it as-is, it should be an impl detail of of the `DialContext`?

And Temporal CLI could also leverage the same `DialContext` internally when it connects to Temporal (as it might also need to refresh the token).

Would you expect the `DialContext` to always include a usable/fresh access token; ie refresh eagerly?